src/GdbCommand.h

```diff
@@ -27,28 +28,146 @@ class GdbCommand {
   virtual std::string invoke(GdbServer& gdb_server, Task* t,
                              const std::vector<std::string>& args) = 0;
 
+  /**
+   * When call, gdb will automatically fill argument to this command
+   * by running gdb.execute(). This is useful to gdb state along side
+   * of the command invocation.
+   */
+  void add_auto_args(const std::string& auto_arg) {
+    cmd_auto_args.push_back(auto_arg);
+  }
+
+  const std::vector<std::string>& auto_args() const { return cmd_auto_args; }
+
+  /**
+   * Setup all the automatic auto_args for our commands.
+   */
+  static void init_auto_args();
+
 private:
   const std::string cmd_name;
+  std::vector<std::string> cmd_auto_args;
 };
 
 #define RR_LINE_CONCAT(str, line) str##line
 #define RR_LINE_EXPAND(str, line) RR_LINE_CONCAT(str, line)
-#define RR_CMD_CLASSNAME() RR_LINE_EXPAND(RRCmd, __LINE__)
 #define RR_CMD_OBJ() RR_LINE_EXPAND(sRRCmdObj, __LINE__)
 
-#define RR_CMD(name)                                                           \
-  class RR_CMD_CLASSNAME() : public GdbCommand {                               \
+#define RR_CMD(id, name)                                                       \
+  class id : public GdbCommand {                                               \
+  public:                                                                      \
+    id() : GdbCommand(name) {}                                                 \
+                                                                               \
+  private:                                                                     \
+    virtual std::string invoke(GdbServer& gdb_server, Task* t,                 \
+                               const std::vector<std::string>& args);          \
```

Comment:
If I understand correctly I think this will make the code worse:
1. Not all commands will need private access to GdbServer. Some commands might just need their private storage tied to the lifetime of GdbServer which we can give them a better way to get that. (I'm not sure if we plan on having multiple GdbServer instances or not).
2. I think this will make readability worse. Having a list of friend class in GdbServer is easy to understand. Its more difficult to understand why we're templating GdbCommandImpl and generating code just to save a few simple repetitive 'friend class' lines.
3. Later if we need to we can define virtual attribute functions for GdbCommand that behave differently in some way. Example could be later add 'virtual bool SupportsDiversion()' and then easily let commands opt into supporting diversion session. With templates we'd have to keep adding more and more to the template arguments and/or use traits.

I don't think making this change is worth it.